hack/reports/profiling.sh

```diff
@@ -0,0 +1,208 @@
+declare -r CPU_PROFILE_DIR="$TMP_DIR/cpu-profile"
+declare -r MEM_PROFILE_DIR="$TMP_DIR/mem-profile"
+declare -r COMPARE_DIR="$TMP_DIR/compare"
+
+source "$PROJECT_ROOT/hack/utils.bash"
+
+validate() {
+	header "Validating"
+	command -v pprof >/dev/null 2>&1 || {
+		fail "No pprof command found in PATH"
+		info "Please install pprof"
+		cat <<-EOF
+			go install github.com/google/pprof@latest
+		EOF
+		# NOTE: do not proceed without pprof installed
+		return 1
+	}
+}
+
+parse_args() {
+	### while there are args parse them
+	while [[ -n "${1+xxx}" ]]; do
+		case $1 in
+		--help | -h)
+			shift
+			SHOW_HELP=true
+			return 0
+			;;
+		--dev-port)
+			shift
+			KEPLER_DEV_PORT="$1"
+			shift
+			;;
+		--latest-port)
+			shift
+			KEPLER_LATEST_PORT="$1"
+			shift
+			;;
+		--duration)
+			shift
+			DURATION="$1"
+			shift
+			;;
+		*) return 1 ;; # show usage on everything else
+		esac
+	done
+	return 0
+}
+
+print_usage() {
+	local scr
+	scr="$(basename "$0")"
+
+	read -r -d '' help <<-EOF_HELP || true
+		  🔆 Usage:
+			  $scr <command> [OPTIONS]
+			  $scr  -h | --help
+
+		    📋 Commands:
+		      capture        Captures CPU and Memory profile from the Kepler dev and latest compose services
+		      compare        Compares two profiles and prints out the top cumulative samples
+
+		    💡 Examples:
+		      →  $scr capture --duration 30 --dev-port 28283 --latest-port 28284
+
+		      →  $scr compare
+
+			⚙️ Options:
+		      -h | --help             Show this help message
+		      --duration              Duration of the profiling session in seconds (default: 30)
+		      --dev-port              Port of the kepler dev compose service  (default: 28283)
+		      --latest-port           Port of the kepler latest compose service (default: 28284)
+	EOF_HELP
+
+	echo -e "$help"
+	return 0
+}
+
+profile_capture() {
+	header "Running CPU Profiling"
+	mkdir -p "$CPU_PROFILE_DIR"
+
+	local http_dev_status=0
+	local http_latest_status=0
+
+	http_dev_status=$(curl -s -o /dev/null -w "%{http_code}" "http://localhost:$KEPLER_DEV_PORT/debug/pprof/")
+	http_latest_status=$(curl -s -o /dev/null -w "%{http_code}" "http://localhost:$KEPLER_LATEST_PORT/debug/pprof/")
+
+	[[ $http_dev_status -ne 200 || $http_latest_status -ne 200 ]] && {
+		fail "Unable to connect to kepler-dev or kepler-latest compose service"
+		info "Make sure that kepler-dev and kepler-latest compose services are running"
+		return 1
+	}
+
+	run pprof -proto -seconds "$DURATION" \
+		-output "$CPU_PROFILE_DIR"/profile-dev.pb.gz \
+		"http://localhost:$KEPLER_DEV_PORT/debug/pprof/profile" &
+
+	run pprof -proto -seconds "$DURATION" \
+		-output "$CPU_PROFILE_DIR"/profile-latest.pb.gz \
+		"http://localhost:$KEPLER_LATEST_PORT/debug/pprof/profile" &
+
+	(
+		sleep 5
+		for i in {1..5}; do
+			info "Querying metric endpoints [$i/5]"
+			curl -s "http://localhost:$KEPLER_DEV_PORT/metrics" >/dev/null &
+			curl -s "http://localhost:$KEPLER_LATEST_PORT/metrics" >/dev/null &
+			wait
+			sleep 5
```

Comment:
lets hit .. 

- latest first 
- dev next
wait 4 seconds 

- dev first
- latest next
wait 4 secs 

So that the distribution is equal between both containers as much as possible